### PR TITLE
✨ SMTP Authentication

### PR DESCRIPTION
**Pull request by cfpwastaken**
Authentication for outgoing SMTP mail (MAIL FROM our host) as described in [RFC 4954](https://datatracker.ietf.org/doc/html/rfc4954) and PLAIN authentication as described in [RFC 4616](https://datatracker.ietf.org/doc/html/rfc4616).

image

### CODE REVIEW

File: src/smtp/SMTPServer.ts [hunk 1]
```diff
@@ -138,11 +147,59 @@
...
+			// The AUTH command is not permitted during a mail transaction.
+			// An AUTH command issued during a mail transaction MUST be rejected with a 503 reply.
+			status(503)
+		}
+
+		if (auth.state == 0) {
+			status(334)
+			auth.state = 1
+		} else if (auth.state == 1) {
+			const [_, username, password] = Buffer.from(msg, "base64").toString().split("\0")
```
[the remark this answers is not in the document]
Comment (cfpwastaken) [reply src/smtp/SMTPServer.ts:174]: Don't think this is getting fixed anytime soon

File: src/smtp/SMTPServer.ts [hunk 2]
```diff
@@ -1,10 +1,11 @@
+import sendStatus, { StatusOptions } from "./status.js"
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:1]:
```suggestion
import sendStatus, { type StatusOptions } from "./status.js"
```

**Review by j0code — CHANGES_REQUESTED** [review]
one small tweak
will approve after

File: src/smtp/SMTPServer.ts [hunk 3]
```diff
@@ -39,6 +40,15 @@ export default class SMTPServer {
 			content: ""
 		}
 
+		// State:
+		// 0: No authentication
+		// 1: Waiting for authentication (AUTH PLAIN)
+		// 2: Authenticated
+		const auth = {
+			state: 0,
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:48]: can you make this an Enum or at least a Union type, e.g. `"off" | "waiting" | "authenticated"`

File: src/smtp/SMTPServer.ts [hunk 4]
```diff
@@ -75,6 +84,18 @@ export default class SMTPServer {
 				}
 				const email = msg.split(":")[1].split(">")[0].replace("<", "")
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:85]:
```suggestion
				const email = msg.substring(msg.indexOf("<") + 1, msg.lastIndexOf(">"))
```
should work
test first though

**Review by j0code — CHANGES_REQUESTED** [review]
minor changes

**Review by j0code — APPROVED** [review]

**Pull request merged**